File: Segundo_Curso/Segundo_Cuatri/PDOO/java/src/irrgarten/FuzzyPlayer.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;
/**
 *
 * @author el_dramas
 */
public class FuzzyPlayer extends Player{
    
    /**
     * Constructor de FuzzyPlayer por copia.
     * @param other Otro jugador.
     */
    public FuzzyPlayer(Player other){
        super(other);
    }
    
    /**
     * Calcula el movimiento de un FuzzyPlayer
     * @param direction Direccion preferente
     * @param validMoves Coleccion de movimientos posibles
     * @return Direccion de movimiento.
     */
    @Override
    public Directions move( Directions direction, ArrayList<Directions> validMoves){
        Directions output = super.move(direction, validMoves);      //Filtramos a direccion para no avanzar a un bloque
        return Dice.nextStep(output, validMoves, this.getIntelligence());
    }
    
    /**
     * Calcula el ataque de un FuzzyPlayer
     * @return Ataque del FuzzyPlayer.
     */
    @Override
    public float attack(){
        return this.sumWeapons()+Dice.intensity(this.getStrength());
    }
    
    /**
     * Calcula la energía defensiva de un FuzzyPlayer.
     * @return Energia defensiva de FuzzyPlayer
     */
    @Override
    protected float defensiveEnergy(){
        return this.sumShield()+Dice.intensity(this.getIntelligence());
    }

    /**
     * Concatena la informacion de una instancia.
     * @return Cadena con la informacion de la instancia.
     */
    @Override
    public String toString(){
        String cad = super.toString();
        cad = cad.replaceFirst("P", "F");
        return cad;
    }
}

File: Segundo_Curso/Segundo_Cuatri/PDOO/ruby/Irrgarten/FuzzyPlayer.rb
require_relative 'Dice'
require_relative 'Player'

module Irrgarten

  class FuzzyPlayer < Player
    @@FUZZY = "Fuzzy"

    # Pseudo-constructor de copia. Como en Ruby ni se pueden
    # sobrecargar es el único que se proporciona
    def initialize(player)
        cloner(player)
    end

    # Se encarga de realizar el movimiento de un FuzzyPlayer
    # Depende del azar y la direccion preferente es filtrada por el metodo move de
    # Player.
    # direction -> direccion preferente
    # validMoves -> array de movimientos validos
    # return -> direccion de movimiento ya calculada
    def move(direction,validMoves)
      direction = Dice.nextStep(super,validMoves,self.intelligence)
      super(direction,validMoves)
    end

    # Ataque de un FuzzyPlayer
    # return -> valor que representa la fuerza de ataque
    def attack
      sumWeapons + Dice.intensity(self.strength)
    end

    # Concatena la informacion de un FuzzyPlayer
    # return -> cadena con la informacion
    def to_s
      @@FUZZY + super
    end

  protected
    # Intensidad defensiva de un FuzzyPlayer
    # return -> Valor que representa la intensidad defensiva.
    def defensiveEnergy
      sumShields + Dice.intensity(self.intelligence)
    end

  end#class
end#module
